torch/autograd/__init__.py

```diff
@@ -265,7 +266,19 @@ def grad(
             cliffs. Please use ``torch._C._debug_only_display_vmap_fallback_warnings(True)``
             to show any performance warnings and file an issue on github if warnings exist
             for your use case. Defaults to ``False``.
+        materialize_grads (bool, optional): If ``True``, set the gradient for unused inputs
+            to zero instead of None. This is useful when computing higher-order derivatives.
+            If ``materialize_grads`` is ``True`` and ``allow_unused`` is ``False``, an error
+            will be raised. Defaults to ``False``.
+
     """
+    if materialize_grads:
+        if allow_unused is False:
+            raise ValueError("Expected allow_unused to be True or not passed when materialize_grads=True,"
+                             "but got: allow_unused=False.")
+        allow_unused = True  # materialize_grads=True implies allow_unused=True
+    if allow_unused is None:
+        allow_unused = False
```

Comment:
I think we can slightly simplify this logic to set allow_unused only once:

```python
if materialize_grads and allow_unused is False:
    raise ValueError("...")
if allow_unused is None:
    allow_unused = materialize_grads
```

Reply:
Added here https://github.com/pytorch/pytorch/pull/97015/commits/105c221556ae5d3b8c37a511d8d10511019de3e3